### PR TITLE
Updates your PR

### PR DESCRIPTION
**Pull request by adam2392**
I created a PR to fix the issues in your PR.

Changes proposed in this pull request:
- UnsupervisedCriterion is now modular and will support TwoMeans and FastBIC probably. TwoMeans only requires a few methods to be overridden.
- UnsupervisedSplitter is implemented
- begin UnsupervisedTree
- [Dependencies] Fix poetry file, so that way you can run `poetry install` and then you will have the relevant dependencies installed. Then you can run the meson build cycle, so that way you can build the Cython files

## Before submitting

- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

**Comment by adam2392** [comment]
@jshinm if you review these changes and merge this in, we can start to do some testing (and fixing any syntax/runtime errors we find along the way) to replicate results of the unsupervised tree paper. 

In parallel, if you get a sense of these changes, you can try implementing the `FastBICCriterion` class

File: sktree/_forest.py [hunk 1]
```diff
@@ -212,3 +213,31 @@ def fit(self, X, y=None, sample_weight=None):
...
+
+    def transform(self, X):
+        from joblib import Parallel
+        from sklearn.utils.fixes import delayed
+
+        results = Parallel(
+            n_jobs=self.n_jobs,
+            verbose=self.verbose,
+            prefer="threads",
+        )(delayed(tree.transform)(X) for tree in self.estimators_)
```
Comment (jshinm) [inline sktree/_forest.py:225]: Note from base.py

        # non-optimized default implementation; override when a better
        # method is possible for a given clustering algorithm

File: sktree/tree/_classes.py [hunk 2]
```diff
@@ -58,4 +134,126 @@ def __init__(
...
+                splitter,
+                min_samples_split,
+                min_samples_leaf,
+                min_weight_leaf,
+                max_depth,
+                max_leaf_nodes,
+                self.min_impurity_decrease,
+            )
+
+        builder.build(self.tree_, X, y, sample_weight)
```
Comment (jshinm) [inline sktree/tree/_classes.py:237]: maybe needs to be changed to take y optional

File: sktree/tree/_unsup_criterion.pxd [hunk 3]
```diff
@@ -39,8 +45,12 @@ cdef class UnsupervisedCriterion(BaseCriterion):
     # Unsupervised criterion can be used with splitter and tree methods.
     cdef int init(
         self,
-        const DTYPE_t[:,:] X,
```
Comment (jshinm) [inline sktree/tree/_unsup_criterion.pxd:42]: init no longer takes in X

File: sktree/tree/_unsup_splitter.pxd [hunk 4]
```diff
@@ -1,93 +1,42 @@
...
-    ) nogil except -1
-
-    cdef void node_value(self, double* dest) nogil
-
-    cdef double node_impurity(self) nogil
-
-cdef class UnsupervisedSplitter(BaseUnsupervisedSplitter):
-    """Abstract interface for supervised splitter."""
+    # feature matrix
+    cdef const DTYPE_t[:, ::1] X
```
Comment (jshinm) [inline sktree/tree/_unsup_splitter.pxd:33]: 2D X, 1D Xf

File: sktree/tree/_unsup_splitter.pyx [hunk 5]
```diff
@@ -216,9 +147,11 @@ cdef class BestSplitter(BaseDenseSplitter):
         cdef DTYPE_t[::1] Xf = self.feature_values
```
Comment (jshinm) [inline sktree/tree/_unsup_splitter.pyx:147]: Xf @ unsup_splitter

File: sktree/tree/_unsup_tree.pxd [hunk 6]
```diff
@@ -84,13 +86,22 @@ cdef class UnsupervisedTreeBuilder:
...
     cdef double min_weight_leaf         # Minimum weight in a leaf
     cdef SIZE_t max_depth               # Maximal tree depth
     cdef double min_impurity_decrease   # Impurity threshold for early stopping
 
-    cpdef build(self, UnsupervisedTree tree, object X, cnp.ndarray sample_weight=*)
-    cdef _check_input(self, object X, cnp.ndarray sample_weight)
+    cpdef build(
+        self,
+        UnsupervisedTree tree,
+        object X,
```
Comment (jshinm) [inline sktree/tree/_unsup_tree.pxd:100]: DTYPE_t ?

File: sktree/tree/_unsup_tree.pxd [hunk 7]
```diff
@@ -84,13 +86,22 @@ cdef class UnsupervisedTreeBuilder:
...
+        cnp.ndarray sample_weight=*
```
Comment (jshinm) [inline sktree/tree/_unsup_tree.pxd:101]: need to review this

File: sktree/_forest.py (hunk 1 above)
Comment (adam2392) [reply sktree/_forest.py:225]: What is this?

File: sktree/tree/_classes.py (hunk 2 above)
Comment (adam2392) [reply sktree/tree/_classes.py:237]: Good catch, `y` shouldn't be even passed. Cython code does not know `y`.

File: sktree/tree/_unsup_criterion.pxd (hunk 3 above)
Comment (adam2392) [reply sktree/tree/_unsup_criterion.pxd:42]: Is this a question, or saying that it's an error?

File: sktree/tree/_unsup_splitter.pxd (hunk 4 above)
Comment (adam2392) [reply sktree/tree/_unsup_splitter.pxd:33]: ?

File: sktree/tree/_unsup_splitter.pyx (hunk 5 above)
Comment (adam2392) [reply sktree/tree/_unsup_splitter.pyx:147]: ?

File: sktree/tree/_unsup_tree.pxd (hunk 6 above)
Comment (adam2392) [reply sktree/tree/_unsup_tree.pxd:100]: Unnecessary. See scikit-learn codebase and https://stackoverflow.com/questions/53938740/difference-between-pyobject-and-object-and-possibly-other-definitions

File: sktree/tree/_unsup_tree.pxd (hunk 7 above)
Comment (adam2392) [reply sktree/tree/_unsup_tree.pxd:101]: http://docs.cython.org/en/latest/src/userguide/pyrex_differences.html#optional-arguments-in-cdef-cpdef-functions

File: sktree/_forest.py (hunk 1 above)
Comment (jshinm) [reply sktree/_forest.py:225]: @adam2392 oh I was just jotting down notes just for my own reference to better understand your PR :laughing: I wasn't at all reviewing your PR. With this particular method, I was looking through how it was being used and I found these comments

File: sktree/tree/_classes.py (hunk 2 above)
Comment (jshinm) [reply sktree/tree/_classes.py:237]: I see, thank you!

File: sktree/tree/_unsup_criterion.pxd (hunk 3 above)
Comment (jshinm) [reply sktree/tree/_unsup_criterion.pxd:42]: I didn't think my comments would ding you when reviews not started! I noted here b/c I think this here was important

File: sktree/tree/_unsup_splitter.pxd (hunk 4 above)
Comment (jshinm) [reply sktree/tree/_unsup_splitter.pxd:33]: I was just reminding myself that X is 2D and Xf is 1D :smile:

File: sktree/tree/_unsup_splitter.pyx (hunk 5 above)
Comment (jshinm) [reply sktree/tree/_unsup_splitter.pyx:147]: Just a reference point to remind myself where Xf is being used

File: sktree/tree/_unsup_tree.pxd (hunk 6 above)
Comment (jshinm) [reply sktree/tree/_unsup_tree.pxd:100]: Thank you!!

File: sktree/tree/_unsup_tree.pxd (hunk 7 above)
Comment (jshinm) [reply sktree/tree/_unsup_tree.pxd:101]: ohh! Thank you :heart:

**Pull request merged**